elf_load.c

```diff
@@ -311,13 +312,224 @@ static int elf_load_imgcpy(struct elf_prog *elf_prog, Elf *elf,
 }
 
 #if CONFIG_LIBVFSCORE
+#if CONFIG_LIBPOSIX_MMAP
+/* If vastart + phdr.p_filesz (vastart) < vastart + phdr.p_memsz (vaend),
+ * 0 out that remainder, either through memset or through anonymous mappings
+ */
+static int elf_load_mmap_filesz_memsz_diff(struct elf_prog *elf_prog, int fd,
+					   GElf_Phdr *phdr __maybe_unused,
+					   uintptr_t vastart, uintptr_t vaend)
+{
+	UK_ASSERT(elf_prog && phdr && vastart && vaend);
+
+	uk_pr_debug("%s: Zeroing 0x%"PRIx64" - 0x%"PRIx64"\n",
+		    elf_prog->name,
+		    (uint64_t)(vastart),
+		    (uint64_t)(vaend));
+
+	/* From the last byte contained in filesz to the last
+	 * byte contained in memsz, we can either have:
+	 * 1. a page alignment adjustment, where the end
+	 * address is just PAGE_ALIGN_UP(paddr + filesz), which
+	 * means that our initial call to mmap already read in
+	 * a page for us whose remaining bytes we memset without
+	 * generating a page fault, because
+	 * vaend - vastart < PAGE_SIZE
+	 * ...
+	 */
+	memset((void *)vastart, 0, PAGE_ALIGN_UP(vastart) - vastart);
+
+	if (vaend == PAGE_ALIGN_UP(vastart))
+		return 0;
+
+	/*
+	 * ...
+	 * 2. vaend - vastart >= PAGE_SIZE, which can happen if
+	 * this segment contains a NOBITS section, such as .bss.
+	 * If that is the case, then simply anonymously map this
+	 * remaining area so that we don't waste time memsetting
+	 * it (.bss is quite large).
+	 */
+	vastart = PAGE_ALIGN_UP(vastart);
+	vastart = (uintptr_t)mmap((void *)vastart, vaend - vastart,
+				  PROT_EXEC | PROT_READ | PROT_WRITE,
+				  MAP_PRIVATE | MAP_FIXED | MAP_ANONYMOUS,
+				  fd, 0);
+	if (unlikely(vastart == (uintptr_t)MAP_FAILED)) {
+		uk_pr_err("Failed to mmap the NOBITS part of phdr at "
+			  "offset %lu\n", phdr->p_offset);
+		return (int)vastart;
+	}
+
+	return 0;
+}
+
+/* Use this to mmap first PT_LOAD */
+static int do_elf_load_fdphdr_0(struct elf_prog *elf_prog,
+				GElf_Phdr *phdr, int fd)
+{
+	uintptr_t vastart_old, vastart_aligned;
+	uintptr_t vaend_old, vaend_new;
+	uintptr_t vastart, vaend;
+	__sz mmap_len;
+	int rc;
+
+	/* First start/vabase can't be !0 before loading first PT_LOAD */
+	UK_ASSERT(!elf_prog->start && !elf_prog->vabase && phdr);
+
+	if (phdr->p_memsz >= phdr->p_align)
+		mmap_len = phdr->p_memsz + phdr->p_align;
```

Comment:
This is the same logic the linker [uses](https://github.com/bminor/glibc/blob/master/elf/dl-map-segments.h#L34) so I thought I would replicate that

Reply:
okey, but is it their mistake? `p_memsz+ p_align` is enough, do you think so?

Reply:
Maybe it is not is `p_memsz < p_align` as the condition implies. :thinking:

Reply:
I know what's my mistake, the alignment needs to be applied to both head and tail, so `p_align * 2` is needed.

```
	/* mmap anonymously what we are left if memsz > filesz */
	vastart = vastart + phdr->p_filesz;
	vaend = PAGE_ALIGN_UP(vastart + (phdr->p_memsz - phdr->p_filesz));
```

But in Line 447, should `vaend` also use `p_align` instead of `PAGE_ALIGN_UP`, for `elf_load_mmap_filesz_memsz_diff` to mmap enough anonymous pages?

Reply:
Isn't `p_align` only for the base address of a segment? Should a segment also care about the alignment of its end address?

Reply:
I'm not sure, have not found the document yet. But if a segment only cares about the alignment of the start address, I still don't understand why `p_align * 2` is needed.

For example, assume memsz is 5KB, and alignment is 1024 KB, so the p_memsz < p_align. If you alloc for a continues 1029KB memory, it is possible that you can't find 5KB memory with 1024KB alignment? Is there an example?

Reply:
Hmm, this commit seems to be the followup of this discussion:
<https://sourceware.org/bugzilla/show_bug.cgi?id=28676>
<https://github.com/bminor/glibc/commit/718fdd87b1b98ef88e883a37d9c18867256fa5a4>

Ok, so the reason it uses this multiply by 2 is because it assumes that the alignment can only be of a certain page size (whether 4K, or 2M, does not matter). So, obviously, if page_size is 2M and you have alignment on 2M and memsz is like 1K, then mmaping 2M + 1K will result in 2M + 2M being mapped anyway.

But, having more context now, I think you're right and it might not have necessarily been handled correctly in the glibc. For example, if you have 4K page size and 2M alignment, then mapping 2M + memsz (which ends up being 2M + 4k for a memsz of 1K), p_align  + memsz is enough. The way it is now 2M + 2M are mapped anyway.

The question is, why do they assume that p_align is always of current system's page size? I think I would go with your suggestion, since it handles both cases :+1:. p_align + p_memsz ends up handling both cases anyway, I like it more. Let's go with that :smile:.

Reply:
I tested with linux, yes, `p_align` is not applied for the tail addresses, page-aligned is enough.